internal/metrics/metrics.go

```diff
@@ -18,41 +18,182 @@ package metrics
+	txHandlerMetricsManager, _ := metricsRegistry.NewMetricsManagerForSubsystem(ctx, metricsTransactionHandlerSubsystemName)
+	_ = metricsRegistry.NewHTTPMetricsInstrumentationsForSubsystem(
+		ctx,
+		metricsRESTAPIServerSubSystemName,
+		true,
+		prometheus.DefBuckets,
+		map[string]string{},
+	)
 	mm := &metricsManager{
-		ctx:            ctx,
-		metricsEnabled: config.GetBool(tmconfig.MetricsEnabled),
-		timeMap:        make(map[string]time.Time),
+		ctx:                     ctx,
+		metricsEnabled:          config.GetBool(tmconfig.MetricsEnabled),
+		timeMap:                 make(map[string]time.Time),
+		metricsRegistry:         metricsRegistry,
+		txHandlerMetricsManager: txHandlerMetricsManager,
 	}
 
 	return mm
 }
 
-func (mm *metricsManager) TransactionSubmissionError() {
-	TransactionSubmissionError.Inc()
+func (mm *metricsManager) IsMetricsEnabled() bool {
+	return mm.metricsEnabled
 }
-func (mm *metricsManager) TransactionsInFlightSet(count float64) {
-	TransactionsInFlight.Set(count)
+
+func (mm *metricsManager) HTTPHandler() http.Handler {
+	httpHandler, _ := mm.metricsRegistry.HTTPHandler(mm.ctx, promhttp.HandlerOpts{})
+	return httpHandler
 }
 
-func (mm *metricsManager) IsMetricsEnabled() bool {
-	return mm.metricsEnabled
+func (mm *metricsManager) GetAPIServerRESTHTTPMiddleware() func(next http.Handler) http.Handler {
+	httpMiddleware, _ := mm.metricsRegistry.GetHTTPMetricsInstrumentationsMiddlewareForSubsystem(mm.ctx, metricsRESTAPIServerSubSystemName)
+	return httpMiddleware
 }
 
 type Metrics interface {
-	TransactionSubmissionError()
-	TransactionsInFlightSet(count float64)
 	IsMetricsEnabled() bool
+
+	// HTTPHandler returns the HTTP handler of this metrics registry
+	HTTPHandler() http.Handler
+
+	GetAPIServerRESTHTTPMiddleware() func(next http.Handler) http.Handler
+
+	// functions for transaction handler to define and emit metrics
+	TransactionHandlerMetrics
+}
+
+// Transaction handler metrics are defined and emitted by transaction handlers
+type TransactionHandlerMetrics interface {
+	// functions for declaring new metrics
+	InitTxHandlerCounterMetric(ctx context.Context, metricName string, helpText string, withDefaultLabels bool)
+	InitTxHandlerCounterMetricWithLabels(ctx context.Context, metricName string, helpText string, labelNames []string, withDefaultLabels bool)
+	InitTxHandlerGaugeMetric(ctx context.Context, metricName string, helpText string, withDefaultLabels bool)
+	InitTxHandlerGaugeMetricWithLabels(ctx context.Context, metricName string, helpText string, labelNames []string, withDefaultLabels bool)
+	InitTxHandlerHistogramMetric(ctx context.Context, metricName string, helpText string, buckets []float64, withDefaultLabels bool)
+	InitTxHandlerHistogramMetricWithLabels(ctx context.Context, metricName string, helpText string, buckets []float64, labelNames []string, withDefaultLabels bool)
+	InitTxHandlerSummaryMetric(ctx context.Context, metricName string, helpText string, withDefaultLabels bool)
```

Comment:
Just a style thing, but what about this to avoid two functions:

```go
InitTxHandlerSummaryMetric(ctx context.Context, metricName string, helpText string, withDefaultLabels bool, extraLabels ...string)
```

Reply:
I looked between optional param vs two separate functions. I picked the latter to inherit the same style from the Prometheus golang library. (`NewCounter` vs `NewCounterVec`).